In [1]:
import pandas as pd
import polars as pl

import numpy as np

import lectura_archivos as fr

# Analisis de Cantidad de Traqueostomias por Servicio de Salud

- En este apartado se quiere ver la cantidad de traqueostomias que se hicieron por Servicio de Salud.
Este analisis se realiza por anio.

## Carga de Archivos de GRD

In [2]:
dfp = fr.leer_grd()

In [4]:
df = dfp.to_pandas()


## Analisis de Traqueostomia

In [5]:
procedimientos = df.loc[:, "PROCEDIMIENTO1":"PROCEDIMIENTO30"]
contiene_nuestros_proc = procedimientos.isin(["31.1", "31.29"]).sum(axis=1).astype(bool)
proc_nacionales = df[contiene_nuestros_proc]

In [6]:
metricas = (
    proc_nacionales.groupby(["ANIO_EGRESO", "SERVICIO_SALUD"])
    .agg(
        n_egresos=("COD_HOSPITAL", "count"),
        peso_medio=("IR_29301_PESO", "mean"),
        estancia_media=("ESTANCIA", "mean"),
    )
    .sort_values(["ANIO_EGRESO", "n_egresos"], ascending=False)
)


In [7]:
metricas.to_excel("output/resumen_traqueostomia_hist.xlsx")


# Analisis de Prestaciones por Diagnostico

- En este apartado se quiere ver la cantidad de prestaciones que se realizan por cada diagnostico
principal.

- El resultado que se quiere obtener es parecido al siguiente:

|Año|Diagnostico|Resumen Procedimientos|
|---|-----------|----------------------|
|2019|"I25.1"|{"45.23": 20, "73.59": 10}|
|2020|"I25.1"|{"45.23": 7, "73.59": 3}|

## Analisis Exploratorio

### Obtener conteo de diagnosticos en el Torax

In [8]:
torax = dfp.filter(pl.col("COD_HOSPITAL") == 112103)
torax.groupby(["DIAGNOSTICO1"]).agg(pl.count()).sort(by="count", descending=True)

shape: (568, 2)
┌──────────────┬───────┐
│ DIAGNOSTICO1 ┆ count │
│ ---          ┆ ---   │
│ str          ┆ u32   │
╞══════════════╪═══════╡
│ I25.1        ┆ 1366  │
│ Z51.1        ┆ 811   │
│ U07.1        ┆ 706   │
│ Z51.4        ┆ 595   │
│ …            ┆ …     │
│ C49.8        ┆ 1     │
│ B67.3        ┆ 1     │
│ Q67.8        ┆ 1     │
│ L89.2        ┆ 1     │
└──────────────┴───────┘

- De estos resultados, se puede ver que el diagnostico Z51.4 es el diagnostico mas frecuente
entre 2019 y 2021. Tuvo 1366 egresos.

- A modo de ejemplo, se quiere ver el dataframe completo del diagnostico Z51.4, y analizar
los procedimientos que tiene incluidos.

### Analisis de conteo de procedimientos

### Forma 1

In [9]:
from functools import reduce

i251 = torax.filter(pl.col("DIAGNOSTICO1") == "I25.1").to_pandas()
listas_procedimientos = (
    i251.loc[:, "PROCEDIMIENTO1":"PROCEDIMIENTO30"].fillna("").agg("-".join, axis=1).str.split("-")
)


sumado = reduce(lambda x, y: x + y, listas_procedimientos.values)


### Forma 2

In [10]:
cambiada = pd.melt(
    i251, id_vars="DIAGNOSTICO1", value_vars=[f"PROCEDIMIENTO{i}" for i in range(1, 31)]
)
cambiada["value"].value_counts()

99.29    1196
99.19    1125
89.54    1035
93.96     930
37.22     806
         ... 
78.61       1
39.98       1
99.10       1
37.99       1
37.87       1
Name: value, Length: 257, dtype: int64

### Analisis Completo

- Se quiere ver la cantida dde procedimientos hechos por posicion de procedimiento. O sea,
en el "PROCEDIMIENTO1" que procedimientos se ingresaron, y cuales son sus frecuencias ordenadas
de mayor a menor?

- Tambien se quiere ver la frecuencia global. O sea, para el diagnostico X, cuales son los
procedimientos que mas se le realizan, independiente de la posicion del procedimiento.

#### Cambio de nombre de columnas PROCEDIMIENTO

- Esto se realiza, ya que permite ordenar estas columnas al utilizar .value_counts() en la
proxima celda.

In [11]:
import string

procedimientos = [f"PROCEDIMIENTO{i}" for i in range(1, 31)]
letras = list(string.ascii_uppercase) + ["a", "b", "c", "d"]
codificacion = dict(zip(procedimientos, letras))
codificacion_reversa = dict(zip(letras, procedimientos))

torax_pandas = torax.to_pandas()
long_prestaciones = pd.melt(
    torax_pandas,
    id_vars=["ANIO_EGRESO", "DIAGNOSTICO1"],
    value_vars=procedimientos,
)

long_prestaciones["variable"] = long_prestaciones["variable"].replace(codificacion)


In [12]:
resumen_prestaciones_global = (
    long_prestaciones.groupby(["ANIO_EGRESO", "DIAGNOSTICO1"])["value"]
    .value_counts()
    .reset_index(name="count")
)

por_procedimiento = (
    long_prestaciones.groupby(["ANIO_EGRESO", "DIAGNOSTICO1", "variable"])["value"]
    .value_counts()
    .reset_index(name="count")
)

por_procedimiento["variable"] = por_procedimiento["variable"].replace(codificacion_reversa)
por_procedimiento = por_procedimiento.rename({"value": "codigo_procedimiento"}, axis=1)

In [13]:
resumen_prestaciones_global.to_excel("output/prestaciones_globales.xlsx")
por_procedimiento.to_excel("output/prestaciones_por_procedimiento.xlsx")


In [23]:
torax.select(
    pl.col('PROCEDIMIENTO1').null_count()
)

shape: (1, 1)
┌────────────────┐
│ PROCEDIMIENTO1 │
│ ---            │
│ u32            │
╞════════════════╡
│ 0              │
└────────────────┘

In [24]:
torax

shape: (13_468, 131)
┌────────────┬────────────┬────────┬────────────┬───┬────────────┬────────────┬────────┬───────────┐
│ COD_HOSPIT ┆ CIP_ENCRIP ┆ SEXO   ┆ FECHA_NACI ┆ … ┆ IR_29301_M ┆ HOSPPROCED ┆ ESTANC ┆ ANIO_EGRE │
│ AL         ┆ TADO       ┆ ---    ┆ MIENTO     ┆   ┆ ORTALIDAD  ┆ ENCIA      ┆ IA     ┆ SO        │
│ ---        ┆ ---        ┆ cat    ┆ ---        ┆   ┆ ---        ┆ ---        ┆ ---    ┆ ---       │
│ i32        ┆ i32        ┆        ┆ date       ┆   ┆ i8         ┆ str        ┆ durati ┆ i32       │
│            ┆            ┆        ┆            ┆   ┆            ┆            ┆ on[ms] ┆           │
╞════════════╪════════════╪════════╪════════════╪═══╪════════════╪════════════╪════════╪═══════════╡
│ 112103     ┆ 1137216    ┆ MUJER  ┆ 1955-10-15 ┆ … ┆ 1          ┆ null       ┆ 3d     ┆ 2019      │
│ 112103     ┆ 1084360    ┆ HOMBRE ┆ 1959-04-03 ┆ … ┆ 3          ┆ null       ┆ 0ms    ┆ 2019      │
│ 112103     ┆ 822388     ┆ MUJER  ┆ 1941-07-20 ┆ … ┆ 3          ┆ null       ┆ 0ms    ┆ 2019      │
│ 112103     ┆ 317615     ┆ HOMBRE ┆ 1949-05-28 ┆ … ┆ 0          ┆ null       ┆ 0ms    ┆ 2019      │
│ …          ┆ …          ┆ …      ┆ …          ┆ … ┆ …          ┆ …          ┆ …      ┆ …         │
│ 112103     ┆ 81034371   ┆ HOMBRE ┆ 1962-01-12 ┆ … ┆ 3          ┆ HOSPITAL   ┆ 27d    ┆ 2021      │
│            ┆            ┆        ┆            ┆   ┆            ┆ CLÍNICO    ┆        ┆           │
│            ┆            ┆        ┆            ┆   ┆            ┆ UNIVERSIDA ┆        ┆           │
│            ┆            ┆        ┆            ┆   ┆            ┆ D DE …     ┆        ┆           │
│ 112103     ┆ 67821076   ┆ HOMBRE ┆ 1962-05-03 ┆ … ┆ 3          ┆ HOSPITAL   ┆ 41d    ┆ 2021      │
│            ┆            ┆        ┆            ┆   ┆            ┆ DR.        ┆        ┆           │
│            ┆            ┆        ┆            ┆   ┆            ┆ ERNESTO    ┆        ┆           │
│            ┆            ┆        ┆            ┆   ┆            ┆ TORRES     ┆        ┆           │
│            ┆            ┆        ┆            ┆   ┆            ┆ GALD…      ┆        ┆           │
│ 112103     ┆ 76198942   ┆ HOMBRE ┆ 1959-10-14 ┆ … ┆ 1          ┆ null       ┆ 4d     ┆ 2021      │
│ 112103     ┆ 73993832   ┆ HOMBRE ┆ 1978-06-17 ┆ … ┆ 2          ┆ null       ┆ 21d    ┆ 2021      │
└────────────┴────────────┴────────┴────────────┴───┴────────────┴────────────┴────────┴───────────┘